tools/torchci/test_insights/file_report_generator.py

```diff
@@ -0,0 +1,690 @@
+
+            has_no_job_name = False
+            for entry in test_data:
+                if "NoJobName" in entry.get("short_job_name", ""):
+                    has_no_job_name = True
+                    break
+            if has_no_job_name:
+                print(f"Has entries with no job name for {head_sha}")
+                continue
+
+            lens.append((head_sha, len(test_data)))
+            del test_data
+
+            if len(lens) > 1:
+                lens.sort(key=lambda x: x[1], reverse=True)
+                if abs(lens[0][1] - lens[1][1]) * 2 / (lens[0][1] + lens[1][1]) < 0.1:
+                    print(f"Using SHA {lens[0][0]} with {lens[0][1]} entries")
+                    return lens[0][0]
+        return None
+
+    def find_suitable_sha(self, date: str) -> Optional[str]:
+        """
+        Auto-select suitable SHA from PyTorch main branch for a given date.
+        Usage:
+        - Provide a date to select a SHA from that day
+        Criteria:
+        - SHA is from main branch
+        - Workflow jobs are successful (green)
+        - S3 test data is available
+        - All test entries have job names
+        """
+
+        print("Searching for suitable SHAs from PyTorch main branch...")
+
+        params = {
+            "start_date": date + " 00:00:00",
+            "stop_date": date + " 23:59:59",
+        }
+
+        # Single query with conditional logic using CASE expressions
+        query = """
+        SELECT
+            w.head_sha,
+            toUnixTimestamp(w.head_commit.'timestamp') as push_date
+        FROM default.workflow_run w
+        WHERE w.head_branch = 'main'
+            AND w.repository.full_name = 'pytorch/pytorch'
+            AND w.name in ('pull', 'trunk', 'inductor', 'slow')
+            AND w.conclusion = 'success'
+            AND w.head_commit.'timestamp' >= {start_date: DateTime}
+            AND w.head_commit.'timestamp' <= {stop_date: DateTime}
+        GROUP BY
+            w.head_sha, w.head_commit.'timestamp'
+        HAVING count(*) >= 4
+        ORDER BY
+            min(w.head_commit.'timestamp') DESC
+        """
+        print(f"Querying ClickHouse for successful shas on {date}")
+        candidates = query_clickhouse(query, params)
+
+        print(f"Found {len(candidates)} candidate SHAs")
+
+        return self._get_first_suitable_sha(candidates)
+
+    @lru_cache
+    def _get_workflow_jobs_for_sha(self, sha: str) -> List[Dict[str, Any]]:
+        """Get workflow runs for a specific SHA using ClickHouse."""
+        query = """
+        SELECT DISTINCT
+            j.id as job_id,
+            j.name as job_name,
+            j.labels as job_labels,
+            j.run_id as workflow_id,
+            j.run_attempt,
+            j.workflow_name as workflow_name
+        FROM default.workflow_job j
+        WHERE j.head_sha = {sha: String}
+            AND j.workflow_name in ('pull', 'trunk', 'inductor', 'slow')
```

Comment:
Same comment about making this parameter configurable on the workflow, i.e. an env variable